### PR TITLE
13020 load schema mapping support

### PR DESCRIPTION
**Pull request by jack-h-wang**
This PR introduces the ability to dynamically configure the location of HL7 to FHIR conversion schemas based on the message type and profile.

This should be reviewed with https://github.com/CDCgov/hl7v2-fhir-converter/pull/8

Test Steps:
1. Run Integration tests and confirm they pass.

## Changes
- Added data structures to support dynamic schema mapping by MSH9/MSH21
- Refactored HL7toFhirTranslator to not rely on a singleton instance
- HL7toFhirTranslator instances can now be constructed with a configuration file path

### Testing
- [x] Tested locally?
- [x] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [x] Added tests?

## Linked Issues
- #13020

### CODE REVIEW

File: prime-router/build.gradle.kts [hunk 1]
```diff
@@ -852,7 +852,7 @@ dependencies {
     implementation("org.yaml:snakeyaml:2.2")
     implementation("io.github.linuxforhealth:hl7v2-fhir-converter") {
         version {
-            branch = "master"
+            branch = "platform/jwang/13020-runtime-constructors"
```
Comment (jack-h-wang) [inline prime-router/build.gradle.kts:855]: Flip this back to master when changes to the HL7 to FHIR converter have been merged

File: prime-router/metadata/test_fhir_mapping/config.properties [hunk 2]
```diff
@@ -0,0 +1,5 @@
+base.path.resource=./metadata/test_fhir_mapping
+supported.hl7.messages=ORU_R01,ORM_O01
+default.zoneid=+00:00
+additional.conceptmap.file=
+additional.resources.location=./metadata/fhir_mapping
```
Comment (jack-h-wang) [inline prime-router/metadata/test_fhir_mapping/config.properties:5]: Test configuration references the original location for additional resources, so none need to be declared here

File: prime-router/metadata/test_fhir_mapping/hl7/message/ORU_R01.yml [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  - resourceName: MessageHeader
+    segment: MSH
+    resourcePath: segments/MSH/MessageHeader
+    repeats: false
+    isReferenced: true
+    additionalSegments:
+      - SFT
+
+
+# This test should only process MSH to produce minimal results
```
Comment (jack-h-wang) [inline prime-router/metadata/test_fhir_mapping/hl7/message/ORU_R01.yml:12]: This only differs from the normal mappings in that MSH is the only segment to be processed, which results in a much smaller output file.

File: prime-router/src/main/kotlin/cli/ProcessFhirCommands.kt [hunk 4]
```diff
@@ -234,7 +234,12 @@ class ProcessFhirCommands : CliktCommand(
             val msh = message.get("MSH") as Segment
             Terser.set(msh, 2, 0, 1, 1, "^~\\&#")
         }
-        return Pair(HL7toFhirTranslator.getInstance().translate(message), message)
+        val hl7profile = HL7Reader.getMessageProfile(message.toString())
```
Comment (jack-h-wang) [inline prime-router/src/main/kotlin/cli/ProcessFhirCommands.kt:237]: Support for additional schema mapping added to CLI commands here

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRConverter.kt [hunk 5]
```diff
@@ -204,7 +204,9 @@ class FHIRConverter(
         // create the hl7 reader
         val hl7Reader = HL7Reader(actionLogger)
         // get the hl7 from the blob store
-        val hl7messages = hl7Reader.getMessages(queueMessage.downloadContent())
+        val hl7rawmessages = queueMessage.downloadContent()
+        val hl7profile = HL7Reader.getMessageProfile(hl7rawmessages)
```
Comment (jack-h-wang) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRConverter.kt:208]: Support for additional schema mapping added to queue processing here

File: prime-router/src/main/kotlin/fhirengine/translation/HL7toFhirTranslator.kt [hunk 6]
```diff
@@ -16,36 +17,16 @@ import org.hl7.fhir.r4.model.Bundle
 /**
  * Translate an HL7 message to FHIR.
  */
-class HL7toFhirTranslator internal constructor(
+class HL7toFhirTranslator(
+    private val configFolderPath: String = "./metadata/fhir_mapping",
```
Comment (jack-h-wang) [inline prime-router/src/main/kotlin/fhirengine/translation/HL7toFhirTranslator.kt:21]: Instead of a singleton instance that is used globally, the caller should now create a HL7toFhirTranslator instance to call the translate function. There should be no change in behavior if constructed without parameters.

File: prime-router/src/main/kotlin/fhirengine/translation/HL7toFhirTranslator.kt [hunk 7]
```diff
@@ -56,7 +37,18 @@ class HL7toFhirTranslator internal constructor(
         hl7Message: Message,
     ): HL7MessageModel {
         val messageTemplateType = getMessageTemplateType(hl7Message)
-        return defaultMessageTemplates[messageTemplateType]
+
+        // if the requested message templates have been previously loaded, return the stored templates.
+        // otherwise, load the templates
+        val messageTemplate = if (messageTemplates.containsKey(configFolderPath)) {
```
Comment (jack-h-wang) [inline prime-router/src/main/kotlin/fhirengine/translation/HL7toFhirTranslator.kt:43]: Templates are stored for performance reasons; otherwise integration test runtime is greatly extended while new instances of ResourceReader load the templates over and over again

File: prime-router/src/main/kotlin/fhirengine/translation/HL7toFhirTranslator.kt [hunk 8]
```diff
@@ -77,6 +69,9 @@ class HL7toFhirTranslator internal constructor(
         // https://github.com/LinuxForHealth/hl7v2-fhir-converter/blob/d5e43fffa96654e7c5bc896e020ff2fa8aac4ff2/src/main/java/io/github/linuxforhealth/hl7/HL7ToFHIRConverter.java#L135-L159
         // If timezone specification is needed it can be provided via a custom HL7MessageEngine with a custom FHIRContext that has the time zone ID set
 
+        // reinitialize the ResourceReader to ensure the correct singleton instance for the job is loaded
+        ResourceReader(ConverterConfiguration(configFolderPath))
```
Comment (jack-h-wang) [inline prime-router/src/main/kotlin/fhirengine/translation/HL7toFhirTranslator.kt:73]: In a number of places processing switches from an local instance of ResourceReader to the singleton; reinitializing here to make sure the two are equivalent.

File: prime-router/src/main/kotlin/fhirengine/utils/HL7Reader.kt [hunk 9]
```diff
@@ -196,6 +196,12 @@ class HL7Reader(private val actionLogger: ActionLogger) : Logging {
     }
 
     companion object {
+        // map of HL7 message profiles: maps name of profile to configuration directory path
+        val messageProfileMap: Map<MessageProfile, String> = emptyMap()
```
Comment (jack-h-wang) [inline prime-router/src/main/kotlin/fhirengine/utils/HL7Reader.kt:200]: The mapping of message profile to configuration path is defined here. This is intentionally left empty for now.

File: prime-router/src/test/kotlin/fhirengine/engine/FhirConverterTests.kt [hunk 10]
```diff
@@ -270,6 +271,36 @@ class FhirConverterTests {
...
+        )
+
+        every { message.downloadContent() }
+            .returns(validHl7)
+
+        val result = engine.getContentFromHL7(message, actionLogger)
+
+        // the test fhir mappings produce far fewer entries than the production ones
+        assertThat(result).isNotEmpty()
+        assertThat(result[0].entry.size).isEqualTo(5)
```
Comment (jack-h-wang) [inline prime-router/src/test/kotlin/fhirengine/engine/FhirConverterTests.kt:301]: Open to suggestions to make this test less brittle. This currently relies on the test mappings producing a specific number of entries.

File: prime-router/src/main/kotlin/fhirengine/translation/HL7toFhirTranslator.kt (hunk 8 above)
Comment (mkalish) [reply prime-router/src/main/kotlin/fhirengine/translation/HL7toFhirTranslator.kt:73]: This seems a bit off?  Is there a unit test that would fail if I uncommented this line out?

File: prime-router/src/main/kotlin/fhirengine/utils/HL7Reader.kt (hunk 9 above)
Comment (mkalish) [inline prime-router/src/main/kotlin/fhirengine/utils/HL7Reader.kt:200]: Nit: maybe update the var name to reflect it's pointing at a directory?

File: prime-router/src/test/kotlin/fhirengine/engine/FhirConverterTests.kt [hunk 11]
```diff
@@ -270,6 +272,38 @@ class FhirConverterTests {
...
+
+        mockkClass(HL7Reader::class)
+        mockkObject(HL7Reader.Companion)
+        every { HL7Reader.Companion.messageProfileMap.get(testProfile) } returns "./metadata/test_fhir_mapping"
+
+        val result2 = engine.getContentFromHL7(message, actionLogger)
+
+        // the test fhir mappings produce far fewer entries than the production ones
+        assertThat(result2).isNotEmpty()
+        assertThat(result2[0].entry.size).isLessThan(result[0].entry.size)
```
Comment (mkalish) [inline prime-router/src/test/kotlin/fhirengine/engine/FhirConverterTests.kt:304]: I think we can go further with these assertions and actually compare the contents

File: prime-router/src/test/kotlin/fhirengine/utils/HL7ReaderTests.kt [hunk 12]
```diff
@@ -271,6 +271,20 @@ OBX|1|test|94558-4^SARS-CoV-2 (COVID-19) Ag [Presence] in Respiratory specimen b
         assertThat(type).isEqualTo("ORU")
     }
 
+    @Test
+    fun `test getMessageProfile`() {
```
Comment (mkalish) [inline prime-router/src/test/kotlin/fhirengine/utils/HL7ReaderTests.kt:275]: Can we add tests to cover cases when profile is null?

**Review by mkalish — CHANGES_REQUESTED** [review]
Really good stuff!  Nothing major from me, just want to see a few more tests.

The only meaningful comment I had was on that initialization of `ResourceReader`

**Review by mkalish — APPROVED** [review]
Test updates look great!
